Task 4/Task 4/Task 4/FolderStateBuilder.cs

```diff
@@ -0,0 +1,170 @@
+                {
+                    subDir.Delete(true);
+                }
+                foreach (FileInfo file in dir.GetFiles())
+                {
+                    file.Delete();
+                }
+            }
+            else
+                throw new IOException("Directopry doesn't exist");
+        }
+
+        private static void BuildFiles(FileEventsInfo file, string directoryPath)
+        {
+            switch (file.EventType)
+            {
+                case FileWatcher.FileActions.Create:
+                    try
+                    {
+                        FileCreation(file);
+                    }
+                    catch
+                    {
+                        CreateSubFolders(file.FullPath);
+                        FileCreation(file);
+                    }
+                    break;
+                case FileWatcher.FileActions.Delete:
+                    File.Delete(file.FullPath);
+                    break;
+                case FileWatcher.FileActions.Change:
+                    try
+                    {
+                        FileChanging(file);
+                    }
+                    catch
+                    {
+                        CreateSubFolders(file.FullPath);
+                        FileChanging(file);
+                    }
+                    break;
+                case FileWatcher.FileActions.Rename:
+                    FileRenaiming(file);
+                    break;
+                default:
+                    break;
+            }
+        }
+
+        /// <summary>
+        /// This method creates files with the help of information from example of
+        /// FileEventsInfo class.
+        /// </summary>
+        private static void FileCreation (FileEventsInfo file)
+        {
+            using (var stream = File.Create(file.FullPath)) { };
+            if (file.Content.Length != 0)
+            {
+                using (StreamWriter writer = new StreamWriter(file.FullPath, true, System.Text.Encoding.UTF8))
+                {
+                    writer.WriteLine(file.Content);
+                }
+            }
+            File.SetCreationTime(file.FullPath, file.LastChangesTime);
+            File.SetLastWriteTime(file.FullPath, file.LastChangesTime);
+        }
+
+        /// <summary>
+        /// This method creates subfolders, if they were created during observation mode.
+        /// It is necessary, because creating empty subfolder is not tracked, and doesn't 
+        /// record in log.
+        /// </summary>
+        private static void CreateSubFolders(string path)
+        {
+            string[] separateString = path.Split('\\');
+            string subFolderPath = String.Empty;
+            for (int i = 0; i < separateString.Length - 1; i++)
+            {
+                subFolderPath += separateString[i];
+                Directory.CreateDirectory(subFolderPath);
+                subFolderPath += @"\";
+            }
+        }
```

Comment:
Docs say that Directory.CreateDirectory creates all subfolders in the path if they weren't created. So you can just pass a full path without filename there (look for a method that removes it in Path class)